### PR TITLE
[Operator] Further performance enhancements for conv2D

### PR DESCRIPTION
**Pull request by Aalanli**
- changed matmul_fp16 to support last dimension a multiple of 2 or 4
- added custom pre-transform kernel to replace auto scheduler generated version
- moved padding to within the conv2d operator to remove redundant padding

### CODE REVIEW

**Comment by Aalanli** [comment]
More changes:
there was a bug in the subgraph rewrite pass that did not include dilations and padding in the new op, therefore the output shapes after the pass are different.

**Comment by yaoyaoding** [comment]
Hi @Aalanli, let me know (by ping me in the PR) if this PR is ready to be reviewed. Sometimes I am not sure whether a PR is ready to be reviewed even when the CI is passed. You can also add [WIP] prefix (work in progress) to the PR title explicitly if you are still working on a PR.

File: python/hidet/graph/ops/conv2d/conv2d.py [hunk 1]
```diff
@@ -39,15 +51,24 @@ def __init__(self, data: TensorNode, weight: TensorNode, stride: List[int], dila
...
+        dtype = data.type.dtype if data.type.dtype.nbytes > weight.type.dtype.nbytes else weight.type.dtype
+
+        def f_compute(wci, kxi, kyi, ni, oci, pi, qi):
+            im_hi = pi * sx + kxi * dilx
+            im_wi = qi * sy + kyi * dily
+            return if_then_else(
+                logical_and(im_hi >= pad_h, im_hi < h + pad_h, im_wi >= pad_w, im_wi < w + pad_w),
+                data[ni, (oci // out_group_size) * wc + wci, im_hi - pad_h, im_wi - pad_w] * weight[oci, wci, kxi, kyi],
+                dtype(pad_value),
+            )
```
Comment (yaoyaoding) [inline python/hidet/graph/ops/conv2d/conv2d.py:64]: We can define a group of prelimiary computation definitions and put it at the "compute ops" (https://github.com/hidet-org/hidet/tree/main/python/hidet/ir/compute/cops) submodule. 

In this case, we can create a submodule hidet.ir.compute.cops.pad to store the computation for padding, and replace the padding computation in pad op and conv2d op.

In the future, when we write a specialized implementation for some operator (like matmul or conv2d), we can reuse the computation definitions.

File: python/hidet/graph/ops/conv2d/conv2d.py [hunk 2]
```diff
@@ -100,13 +121,28 @@ def __init__(self, data: TensorNode, weight: TensorNode, stride: List[int], dila
...
+    ):
         stride = normalize_stride(stride)
         dilations = normalize_dilations(dilations)
         super().__init__(
             inputs=[x, w],
-            attributes={'stride': stride, 'groups': groups, 'dilations': dilations},
-            task=Conv2dTask(input_like(x, 'x'), input_like(w, 'w'), stride, dilations, groups),
+            attributes={
+                'padding': padding,
+                'pad_value': pad_value,
```
Comment (yaoyaoding) [inline python/hidet/graph/ops/conv2d/conv2d.py:140]: 1. Is this attribute necessary? (I mean any operator in onnx/pytorch used it?). 
2. Better move this attribute to the end of the attribute list, as it is the least used one.

File: python/hidet/graph/ops/matmul/matmul_f16.py [hunk 3]
```diff
@@ -329,9 +360,9 @@ def matmul_f16(a: Tensor, b: Tensor, parallel_k_parts=1) -> Tensor:
         raise ValueError('a and b must have at least 2 dimensions, got shape {} and {}'.format(a.shape, b.shape))
     # TODO: impliment dynamic run-time shape assertion
     if not (isinstance(a.shape[-1], Expr) or isinstance(b.shape[-1], Expr)) and (
-        a.shape[-1] % 8 != 0 or b.shape[-1] % 8 != 0
+        a.shape[-1] % 2 != 0 or b.shape[-1] % 2 != 0
     ):
-        raise ValueError('Expect the last dimension of the input tensors to be a multiple of 8')
+        raise ValueError('Expect the last dimension of the input tensors to be a multiple of 8 or 4')
```
Comment (yaoyaoding) [inline python/hidet/graph/ops/matmul/matmul_f16.py:365]:
```suggestion
        raise ValueError('Expect the last dimension of the input tensors to be a multiple of 2')
```

File: tests/operators/test_conv2d.py [hunk 4]
```diff
@@ -122,15 +125,40 @@ def test_conv2d_channel_last(n, c, h, w, oc, kx, ky, groups, stride, dilations):
...
+        [1, 3, 32, 32, 12, 3, 3],  # kernel 3,
+        [2, 3, 32, 32, 12, 7, 7],  # kernel 7, batch size 2
+        [1, 3, 32, 32, 12, 1, 1],  # kernel 1,
+    ],
+)
+@pytest.mark.parametrize("padding", [[0, 0], [1, 2]])
+@pytest.mark.parametrize("stride", [[1, 1], [2, 3]])
+@pytest.mark.parametrize("dilations", [[1, 1], [2, 3]])
+def test_conv2d_gemm(n, c, h, w, oc, kx, ky, padding, stride, dilations):
+    check_binary(
```
Comment (yaoyaoding) [inline tests/operators/test_conv2d.py:155]: Let's reduce the number of tests here. The previous PR (https://github.com/hidet-org/hidet/pull/283/files) adds 10 more minutes. We can  use the configs like
```
    "n, c, h, w, oc, kx, ky,padding,stride,dilations",
    [
        [1, 3, 32, 32, 12, 3, 3, [0, 0], [1, 1], [1, 1]],  # kernel 3,
        [2, 3, 32, 32, 12, 7, 7, [1, 2], [2, 3], [2, 3]],  # kernel 7, batch size 2
        [1, 3, 32, 32, 12, 1, 1, [0, 0], [2, 3], [1, 1]],  # kernel 1,
    ],
```

File: tests/operators/test_conv2d_pretransform.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        img = hidet.ops.pad(img, [0, pad_channel])
+    return img
+
+
+@pytest.mark.skip(reason='This operator is not needed right now')
+@pytest.mark.parametrize("img_dim", [[32, 64], [31, 63]])
+@pytest.mark.parametrize("channel", [3, 32, 64])
+@pytest.mark.parametrize("padding", [[0, 0], [1, 1], [2, 3]])
+@pytest.mark.parametrize("multi_8", [True, False])
+def test_pretransform_v3(img_dim, channel, padding, multi_8):
```
Comment (yaoyaoding) [inline tests/operators/test_conv2d_pretransform.py:29]: Move to test_conv2d?

**Review by yaoyaoding — COMMENTED** [review]
Thanks @Aalanli, I left some comments.

File: python/hidet/graph/ops/conv2d/conv2d.py (hunk 2 above)
Comment (Aalanli) [reply python/hidet/graph/ops/conv2d/conv2d.py:140]: torch supports padding mode, so this is not exactly the same. Removing for now.

**Review by yaoyaoding — APPROVED** [review]
Looks good to me. Thanks @Aalanli !

**Pull request merged**